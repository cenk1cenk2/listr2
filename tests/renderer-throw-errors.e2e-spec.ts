import delay from 'delay'
import { mockProcessExit, mockProcessStderr, mockProcessStdout } from 'jest-mock-process'

import { Listr } from '@root/index'

describe('show throw error', () => {
  let mockExit: jest.SpyInstance<never, [number?]>

  // eslint-disable-next-line @typescript-eslint/ban-types
  let mockStdout: jest.SpyInstance<boolean, [string, string?, Function?]>

  // eslint-disable-next-line @typescript-eslint/ban-types
  let mockStderr: jest.SpyInstance<boolean, [string, string?, Function?]>

  process.stdout.isTTY = true

  beforeEach(async () => {
    mockExit = mockProcessExit()
    mockStdout = mockProcessStdout()
    mockStderr = mockProcessStderr()
  })

  afterEach(async () => {
    mockExit.mockRestore()
    mockStdout.mockRestore()
    mockStderr.mockRestore()
    jest.clearAllMocks()
  })

  // vxSr6uKuGeO4iX7r2j2Uo9SYlXRhz6nf
  it('should throw out an error', async () => {
    let err: Error | undefined
    try {
      await new Listr(
        [
          {
            title: 'This task will fail.',
            task: async (): Promise<void> => {
              await delay(20)
              throw new Error('This task failed after 2 seconds.')
            }
          },
          {
            title: 'This task will never execute.',
            task: (ctx, task): void => {
              task.title = 'I will change my title if this executes.'
            }
          }
        ],
        {
          concurrent: false,
          exitOnError: true,
          rendererOptions: { lazy: true }
        }
      ).run()
    } catch (e) {
      err = e
    }

    expect(err).toBeTruthy()
    expect(mockStdout.mock.calls).toMatchSnapshot('vxSr6uKuGeO4iX7r2j2Uo9SYlXRhz6nf-out')
    expect(mockStderr.mock.calls).toMatchSnapshot('vxSr6uKuGeO4iX7r2j2Uo9SYlXRhz6nf-err')
    expect(mockExit.mock.calls).toMatchSnapshot('vxSr6uKuGeO4iX7r2j2Uo9SYlXRhz6nf-exit')
  })

  // OgQG0pbo8qgn5ZD8TLdlzM3HnnnYoozl
  it('should not throw out an error when exitonerror is disabled', async () => {
    let err: Error | undefined
    try {
      await new Listr(
        [
          {
            title: 'This task will fail.',
            task: async (): Promise<void> => {
              await delay(20)
              throw new Error('This task failed after 2 seconds.')
            }
          },
          {
            title: 'This task will execute.',
            task: (ctx, task): void => {
              task.title = 'I will change my title if this executes.'
            }
          }
        ],
        {
          concurrent: false,
          exitOnError: false,
          rendererOptions: { lazy: true }
        }
      ).run()
    } catch (e) {
      err = e
    }

    expect(err).toBeFalsy()
    expect(mockStdout.mock.calls).toMatchSnapshot('OgQG0pbo8qgn5ZD8TLdlzM3HnnnYoozl-out')
    expect(mockStderr.mock.calls).toMatchSnapshot('OgQG0pbo8qgn5ZD8TLdlzM3HnnnYoozl-err')
    expect(mockExit.mock.calls).toMatchSnapshot('OgQG0pbo8qgn5ZD8TLdlzM3HnnnYoozl-exit')
  })

  // HUHzWgxW8O1FGgEisITyBb2eDKzmyhUO
  it('should change exit on error per subtask', async () => {
    let err: Error | undefined
    try {
      await new Listr(
        [
          {
            title: 'This task will execute and not quit on errors.',
            task: (ctx, task): Listr =>
              task.newListr(
                [
                  {
                    title: 'This is a subtask.',
                    task: async (): Promise<void> => {
                      throw new Error('I have failed [0]')
                    }
                  },
                  {
                    title: 'This is an another subtask.',
                    task: async (): Promise<void> => {
                      throw new Error('I have failed [1]')
                    }
                  },
                  {
                    title: 'This is yet an another subtask.',
                    task: async (ctx, task): Promise<void> => {
                      task.title = 'I have succeeded.'
                    }
                  }
                ],
                { exitOnError: false }
              )
          },
          {
            title: 'This task will execute.',
            task: (): void => {
              throw new Error('I will exit on error since I am a direct child of parent task.')
            }
          }
        ],
        {
          concurrent: false,
          exitOnError: true,
          rendererOptions: { lazy: true }
        }
      ).run()
    } catch (e) {
      err = e
    }

    expect(err).toBeTruthy()
    expect(mockStdout.mock.calls).toMatchSnapshot('HUHzWgxW8O1FGgEisITyBb2eDKzmyhUO-out')
    expect(mockStderr.mock.calls).toMatchSnapshot('HUHzWgxW8O1FGgEisITyBb2eDKzmyhUO-err')
    expect(mockExit.mock.calls).toMatchSnapshot('HUHzWgxW8O1FGgEisITyBb2eDKzmyhUO-exit')
  })

  // H2KTg7q5F1kWMtrPFdOERVSZc3UT2IsM
  it('should contain errors from failed tasks', async () => {
    let err: Error | undefined
    const task: Listr<any, any, any> = new Listr(
      [
        {
          title: 'This task will execute and not quit on errors.',
          task: (ctx, task): Listr =>
            task.newListr(
              [
                {
                  title: 'This is a subtask.',
                  task: async (): Promise<void> => {
                    throw new Error('I have failed [0]')
                  }
                },
                {
                  title: 'This is an another subtask.',
                  task: async (): Promise<void> => {
                    throw new Error('I have failed [1]')
                  }
                },
                {
                  title: 'This is yet an another subtask.',
                  task: async (ctx, task): Promise<void> => {
                    task.title = 'I have succeeded.'
                  }
                }
              ],
              { exitOnError: false }
            )
        },
        {
          title: 'This task will execute.',
          task: (): void => {
            throw new Error('I will exit on error since I am a direct child of parent task.')
          }
        }
      ],
      {
        concurrent: false,
        exitOnError: true,
        rendererOptions: { lazy: true }
      }
    )

    try {
      await task.run()
    } catch (e) {
      err = e
    }

    expect(err).toBeTruthy()
    expect(task.err).toMatchSnapshot('H2KTg7q5F1kWMtrPFdOERVSZc3UT2IsM')
  })

  // 03IyrStkPGQBIcbYM0HQXoYQxDEVZu8H
  it('should throw out an error in the data field with collapse errors false', async () => {
    let err: Error | undefined
    try {
      await new Listr(
        [
          {
            title: 'This task will fail.',
            task: async (): Promise<void> => {
              await delay(20)
              throw new Error('This task failed after 2 seconds.')
            }
          },
          {
            title: 'This task will never execute.',
            task: (ctx, task): void => {
              task.title = 'I will change my title if this executes.'
            }
          }
        ],
        {
          concurrent: false,
          exitOnError: true,
          rendererOptions: { lazy: true, collapseErrors: false }
        }
      ).run()
    } catch (e) {
      err = e
    }

    expect(err).toBeTruthy()
    expect(mockStdout.mock.calls).toMatchSnapshot('03IyrStkPGQBIcbYM0HQXoYQxDEVZu8H-out')
    expect(mockStderr.mock.calls).toMatchSnapshot('03IyrStkPGQBIcbYM0HQXoYQxDEVZu8H-err')
    expect(mockExit.mock.calls).toMatchSnapshot('03IyrStkPGQBIcbYM0HQXoYQxDEVZu8H-exit')
  })

  // lnjpjmnHOxRSKy9J6YCMtqSAsVkHC3mH
  it('should show the default task title when failed with showErrorMessage of', async () => {
    let err: Error | undefined
    try {
      await new Listr(
        [
          {
            title: 'This task will fail.',
            task: async (): Promise<void> => {
              await delay(20)
              throw new Error('This task failed after 2 seconds.')
            }
          },
          {
            title: 'This task will never execute.',
            task: (ctx, task): void => {
              task.title = 'I will change my title if this executes.'
            }
          }
        ],
        {
          concurrent: false,
          exitOnError: true,
          rendererOptions: { lazy: true, showErrorMessage: false }
        }
      ).run()
    } catch (e) {
      err = e
    }

    expect(err).toBeTruthy()
    expect(mockStdout.mock.calls).toMatchSnapshot('lnjpjmnHOxRSKy9J6YCMtqSAsVkHC3mH-out')
    expect(mockStderr.mock.calls).toMatchSnapshot('lnjpjmnHOxRSKy9J6YCMtqSAsVkHC3mH-err')
    expect(mockExit.mock.calls).toMatchSnapshot('lnjpjmnHOxRSKy9J6YCMtqSAsVkHC3mH-exit')
  })
})
