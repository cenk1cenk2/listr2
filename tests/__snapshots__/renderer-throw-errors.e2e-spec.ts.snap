// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`show throw error should change exit on error per subtask: HUHzWgxW8O1FGgEisITyBb2eDKzmyhUO-err 1`] = `Array []`;

exports[`show throw error should change exit on error per subtask: HUHzWgxW8O1FGgEisITyBb2eDKzmyhUO-exit 1`] = `Array []`;

exports[`show throw error should change exit on error per subtask: HUHzWgxW8O1FGgEisITyBb2eDKzmyhUO-out 1`] = `Array []`;

exports[`show throw error should contain errors from failed tasks: H2KTg7q5F1kWMtrPFdOERVSZc3UT2IsM 1`] = `
Array [
  [ListrError: I will exit on error since I am a direct child of parent task.],
]
`;

exports[`show throw error should not throw out an error when exitonerror is disabled: OgQG0pbo8qgn5ZD8TLdlzM3HnnnYoozl-err 1`] = `Array []`;

exports[`show throw error should not throw out an error when exitonerror is disabled: OgQG0pbo8qgn5ZD8TLdlzM3HnnnYoozl-exit 1`] = `Array []`;

exports[`show throw error should not throw out an error when exitonerror is disabled: OgQG0pbo8qgn5ZD8TLdlzM3HnnnYoozl-out 1`] = `Array []`;

exports[`show throw error should show the default task title when failed with showErrorMessage of: lnjpjmnHOxRSKy9J6YCMtqSAsVkHC3mH-err 1`] = `Array []`;

exports[`show throw error should show the default task title when failed with showErrorMessage of: lnjpjmnHOxRSKy9J6YCMtqSAsVkHC3mH-exit 1`] = `Array []`;

exports[`show throw error should show the default task title when failed with showErrorMessage of: lnjpjmnHOxRSKy9J6YCMtqSAsVkHC3mH-out 1`] = `Array []`;

exports[`show throw error should throw out an error in the data field with collapse errors false: 03IyrStkPGQBIcbYM0HQXoYQxDEVZu8H-err 1`] = `Array []`;

exports[`show throw error should throw out an error in the data field with collapse errors false: 03IyrStkPGQBIcbYM0HQXoYQxDEVZu8H-exit 1`] = `Array []`;

exports[`show throw error should throw out an error in the data field with collapse errors false: 03IyrStkPGQBIcbYM0HQXoYQxDEVZu8H-out 1`] = `Array []`;

exports[`show throw error should throw out an error: vxSr6uKuGeO4iX7r2j2Uo9SYlXRhz6nf-err 1`] = `
Array [
  Array [
    "[?25l",
  ],
  Array [
    "[?25h",
  ],
]
`;

exports[`show throw error should throw out an error: vxSr6uKuGeO4iX7r2j2Uo9SYlXRhz6nf-exit 1`] = `Array []`;

exports[`show throw error should throw out an error: vxSr6uKuGeO4iX7r2j2Uo9SYlXRhz6nf-out 1`] = `
Array [
  Array [
    "[?25l",
  ],
  Array [
    "‚óº This task will fail.
",
  ],
  Array [
    "[2K[1A[2K[G",
  ],
  Array [
    "‚óº This task will fail.
‚óº This task will never execute.
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G",
  ],
  Array [
    "‚óº This task will fail.
‚óº This task will never execute.
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G",
  ],
  Array [
    "‚óº This task will fail.
‚óº This task will never execute.
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G",
  ],
  Array [
    "‚ùØ This task will fail.
‚óº This task will never execute.
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G",
  ],
  Array [
    "‚úñ This task will fail.
‚óº This task will never execute.
",
  ],
  Array [
    "[2K[1A[2K[G",
  ],
  Array [
    "‚óº This task will never execute.
",
  ],
  Array [
    "[2K[1A[2K[G",
  ],
  Array [
    "‚óº This task will never execute.
",
  ],
  Array [
    "[?25h",
  ],
]
`;
